### ISSUE TITLE
Opcode to do indexed array access without a constant table entry

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#296]
Currently code like `arr[123]` generates a `GETPROP` with a numeric constant in the constant table. When the key is a small non-negative integer, could use a specific opcode `GETIDX` which might be slightly faster.

**Comment by fatcerberus** [comment]
Hm, I'm not sure how much of a performance gain this will be in real-world code.  Typically arrays are iterated through with a for loop, thus indexing is done through a variable which gets no gain from this.  I don't really see that much direct indexing.

**Comment by svaarala** [comment]
Sometimes arrays are used to represent tuples of values or fixed-width elements like coordinates or similar. I agree it's not very common.

**Comment by reqshark** [comment]
i never use for loops.

just decrement a value given for evaluation by a while loop, when it gets to zero it stops, less overhead.

**Comment by fatcerberus** [comment]
Yeah, same principle though--the array is indexed through a variable.  The suggested change is for an opcode to index directly using a constant index.

**Comment by reqshark** [comment]
high performance code is generally uncommon.

while decrementation is a virtue. i was discussing a different principle, the virtue of performance

**Comment by creationix** [comment]
I would use arrays as tuples with constant indexes if there was a good fast path for it.  I've been writing a lot of lua lately and love the multiple return values in certain use cases.  The JS version would probably use tiny arrays. 

Could we even go as far as create a special tuple type with fixed length (for code like `[1,2,3]`) and upgrade it to a real array if something happens that violates that constraint?  Or would that add too much code?

**Comment by fatcerberus** [comment]
Hm, that's true, I hadn't thought of tuples, but that's a good use case.  I probably do the same thing without even thinking about it, now that I think of it.

**Comment by svaarala** [comment]
@creationix This issue is tracking a pretty low level bytecode executor improvement. It's not enough to make multireturns (which are provided by ES6 using destructuring assignment) efficient. In particular, for proper multireturns one would expect to avoid creation of an actual array object at all when handling the multireturn.

I wrote a little bit about that into a separate issue: #356.

In short, much of the destructuring/multireturn feature would be implementable now, but a few crucial bits depend on reworking the compiler to have an intermediate representation. I don't think trying to hack the current compiler to do this would work well (at least that's my current throughts on the matter).

**Comment by creationix** [comment]
Sounds good, destructuring would be prettier code anyway.  That's essentially my use case.

**Comment by svaarala** [comment]
Yeah, that's one of the big things I'm looking forward to from ES6 :)

**Comment by fatcerberus** [comment]
Destructuring is indeed nice, but personally I'm holding out for fat arrow functions.

**Comment by creationix** [comment]
I wish es6 had conditional catch from spidermonkey.  That would really cleanup my dukluv error handling code.

```
[catch (exception_var_1 if condition_1) { // non-standard
   catch_statements_1
}]
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch

@svaarala would you condifer adding that even though it's not slated for ES6?

**Comment by fatcerberus** [comment]
I feel like that would be a nightmare to implement and confusing to follow as well.  It reminds me a lot CoffeeScript idioms though, so I'm probably biased (I hate CS).
